File: Algorithms/1008_Construct_Binary_Search_Tree_from_Preorder_Traversal/Java/ConstructBinarySearchTreefromPreorderTraversalSolution1.java

// Space: O(1)
// Time: O(n)

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public TreeNode bstFromPreorder(int[] preorder) {
        if (preorder.length == 0) return null;
        if (preorder.length == 1) return new TreeNode(preorder[0]);
        TreeNode root = new TreeNode(preorder[0]);
        int index = 1;
        int [] left , right;
        while (index < preorder.length && preorder[index]<preorder[0]) {
                index++;
        }
        left = Arrays.copyOfRange(preorder,1,index);
        right = Arrays.copyOfRange(preorder,index,preorder.length);
        root.left = bstFromPreorder(left);
        root.right = bstFromPreorder(right);
        return root;
    }
}

File: Algorithms/1008_Construct_Binary_Search_Tree_from_Preorder_Traversal/Python/Construct_Binary_Search_Tree_from_Preorder_Traversal_Solution_1.py

# Space: O(1)
# Time: O(n)

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def bstFromPreorder(self, preorder):
        return self.dfs(preorder)

    def dfs(self, alist):
        if len(alist) == 0: return None
        if len(alist) == 1: return TreeNode(alist[0])
        root = TreeNode(alist[0])
        index = 1
        left, right = [], []
        while index < len(alist):
            if alist[index] < alist[0]:
                left.append(alist[index])
                index += 1
            else:
                break
        right = alist[index:]
        root.left = self.dfs(left)
        root.right = self.dfs(right)
        return root




